components/kyma-environment-broker/internal/director/client.go

```diff
@@ -1,33 +1,161 @@
+	Result graphql.RuntimeExt `json:"result"`
+}
+
+var lock sync.Mutex
+
+// NewDirectorClient returns new director client struct pointer
+func NewDirectorClient(oauthClient OauthClient, gqlClient GraphQLClient) *Client {
+	return &Client{
+		graphQLClient: gqlClient,
+		oauthClient:   oauthClient,
+		queryProvider: queryProvider{},
+		token:         oauth.Token{},
 	}
 }
 
-// TODO: wrap interface
-type directorClient struct {
-	oauthClient oauth.Client
-	token       oauth.Token
+// GetConsoleURL fetches, validates and returns console URL from director component based on runtime ID
+func (dc *Client) GetConsoleURL(accountID, runtimeID string) (string, error) {
+	log.Info("DirectorClient: Create request to director service")
+	query := dc.queryProvider.Runtime(runtimeID)
+	req := machineGraph.NewRequest(query)
+	req.Header.Add(accountIDKey, accountID)
+
+	log.Info("DirectorClient: Send request to director")
+	response, err := dc.callDirector(req)
+	if err != nil {
+		// do not wrap error, because type of error (TemporaryError) is important
+		return "", err
+	}
+
+	log.Info("DirectorClient: Extract the URL from the response")
+	return dc.getURLFromRuntime(&response.Result)
 }
 
-func (cc *directorClient) setToken() error {
-	token, err := cc.oauthClient.GetAuthorizationToken()
+func (dc *Client) callDirector(req *machineGraph.Request) (*successResponse, error) {
+	var response *successResponse
+	var lastError error
+	var success bool
+	authorizationKey := "Authorization"
+
+	for i := 0; i < reqAttempt; i++ {
+		err := dc.setToken()
+		if err != nil {
+			lastError = err
+			log.Errorf("cannot set token to director client (attempt %d): %s", i, err)
+			continue
+		}
+		req.Header.Add(authorizationKey, fmt.Sprintf("Bearer %s", dc.token.AccessToken))
+		response, err = dc.call(req)
+		if err != nil {
+			lastError = err
+			dc.token.AccessToken = ""
+			req.Header.Del(authorizationKey)
+			log.Errorf("call to director failed (attempt %d): %s", i, err)
+			continue
+		}
+		success = true
+		break
+	}
+
+	if !success {
+		return &successResponse{}, lastError
+	}
+
+	return response, nil
+}
+
+func (dc *Client) call(req *machineGraph.Request) (*successResponse, error) {
+	var response successResponse
+	err := dc.graphQLClient.Run(context.Background(), req, &response)
 	if err != nil {
-		return errors.Wrap(err, "Error while obtaining token")
+		return nil, TemporaryError{fmt.Sprintf("while requesting to director client: %s", err)}
 	}
+	return &response, nil
+}
 
-	if token.EmptyOrExpired() {
-		return errors.New("Obtained empty or expired token")
+func (dc *Client) setToken() error {
+	lock.Lock()
+	defer lock.Unlock()
+	if !dc.token.EmptyOrExpired() {
+		return nil
 	}
 
-	cc.token = token
+	token, err := dc.oauthClient.GetAuthorizationToken()
+	if err != nil {
+		return errors.Wrap(err, "Error while obtaining token")
+	}
+	dc.token = token
+
 	return nil
 }
+
+func (dc Client) getURLFromRuntime(response *graphql.RuntimeExt) (string, error) {
```

Comment:
use pointer receiver

Reply:
I changed but IMO it is not necessary, inside method any of client field is not changed.